ir/src/depgraph.rs

```diff
@@ -0,0 +1,433 @@
+use filetime::FileTime;
+use serde::{Deserialize, Serialize};
+
+use std::{
+    collections::{HashMap, HashSet},
+    fs, io,
+    path::{Path, PathBuf},
+    vec,
+};
+
+/// Helps identify changes in a set of files.
+#[derive(Serialize, Deserialize, Debug, Clone)]
+#[serde(from = "DepGraphSerdeRepr", into = "DepGraphSerdeRepr")]
+pub struct DepGraph {
+    entries: HashMap<PathBuf, Option<DepGraphEntry>>,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+struct DepGraphEntry {
```

Comment:
can we think of a better name here? `FileState`? `FileInfo`? `FileStat`?

Reply:
Also (and just thinking out loud, not a suggestion per se) instead of wrapping this in an `Option` it could be its own enum, like,

```rust
pub enum FileStat {
    Seen { mtime: FileTime, byte_size: u64 },
    Unknown,
}
```

which might make some of your other logic easier? Or not 💁

Reply:
I quite like the enum idea, will try it out

Reply:
*much* better